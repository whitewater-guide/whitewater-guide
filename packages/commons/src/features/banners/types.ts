import { Connection, NamedNode, Node, NodeRef } from '../../apollo';
import { Group } from '../groups';
import { Region } from '../regions';

export enum BannerPlacement {
  MOBILE_SECTION_DESCRIPTION = 'mobileSectionDescription',
  MOBILE_SECTION_ROW = 'mobileSectionRow',
  MOBILE_SECTION_MEDIA = 'mobileSectionMedia',
  MOBILE_REGION_DESCRIPTION = 'mobileRegionDescription',
}

export enum BannerKind {
  Image = 'Image',
  WebView = 'WebView',
}

export interface BannerSource {
  kind: BannerKind;
  ratio: number | null;
  src: string;
}

export interface BannerCore extends NamedNode {
  enabled: boolean;
  slug: string;
  priority: number;
  placement: BannerPlacement;
  source: BannerSource;
  link: string | null;
  extras: any;
}

export interface Banner extends BannerCore {
  // --- connections
  regions?: Connection<Region>;
  groups?: Connection<Group>;
  // ---- transient
  deleted?: boolean;
}

export interface BannerSourceInput {
  kind: BannerKind;
  ratio: number | null;
  src: string | null;
}

export interface BannerInput {
  id: string | null;
  slug: string;
  name: string;
  enabled: boolean;
  priority: number;
  placement: BannerPlacement;
  source: BannerSourceInput;
  link: string | null;
  extras: { [key: string]: any } | null;
  // --- connections
  regions: NodeRef[];
  groups: NodeRef[];
}

export function isBanner(node: Node): node is Banner {
  return node.__typename === 'Banner';
}
