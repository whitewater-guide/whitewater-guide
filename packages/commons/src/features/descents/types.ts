import { Timestamped, PageInfo } from '../../apollo';
import { Section } from '../sections';
import { User } from '../users';

export interface DescentLevel {
  __typename?: 'Level';
  unit?: string | null;
  value?: number | null;
}

export interface DescentLevelInput {
  unit?: string | null;
  value: number;
}

export interface Descent extends Timestamped {
  __typename?: 'Descent';
  id: string;
  user: User;
  section: Section;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevel | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentEdge<TCursor = any> {
  __typename?: 'DescentEdge';
  node: Descent;
  cursor: TCursor;
}

export interface DescentInput {
  id?: string | null;

  sectionId: string;
  startedAt: string;
  duration?: number | null;
  level?: DescentLevelInput | null;
  comment?: string | null;
  public?: boolean | null;
}

export interface DescentsConnection<TCursor = any> {
  __typename?: 'DescentsConnection';
  edges: DescentEdge[];
  pageInfo: PageInfo<TCursor>;
}

export interface DescentsFilter {
  startDate?: string | null;
  endDate?: string | null;
  difficulty?: [number, number] | null;
  userId?: string | null;
  sectionId?: string | null;
  sectionName?: string | null;
}
