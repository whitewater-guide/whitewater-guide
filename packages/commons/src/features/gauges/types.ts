import { NamedNode, Node, Timestamped } from '../../core';
import { HarvestStatus, Measurement } from '../measurements';
import { Point, PointInput } from '../points';
import { Source } from '../sources';

export interface Gauge extends NamedNode, Timestamped {
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
  enabled: boolean;
  location: Point;
  source: Source;
  // measurements?: Measurement[];
  lastMeasurement: Measurement | null;
  status: HarvestStatus | null;
}

export interface GaugeInput {
  id: string | null;
  name: string;
  source: Node;
  location: PointInput | null;
  code: string;
  levelUnit: string | null;
  flowUnit: string | null;
  requestParams: any;
  cron: string | null;
  url: string | null;
}
