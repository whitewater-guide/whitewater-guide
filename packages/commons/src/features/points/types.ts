import { Node } from '../../apollo';

export type Coordinate2d = [number, number];
export type Coordinate3d = [number, number, number];
export type Coordinate = Coordinate2d | Coordinate3d;

export interface Point extends Node {
  name: string | null;
  description: string | null;
  coordinates: Coordinate3d;
  kind: string;
}

export interface PointInput {
  id: string | null;
  name?: string | null;
  description?: string | null;
  coordinates: Coordinate3d | null;
  kind: string;
}
