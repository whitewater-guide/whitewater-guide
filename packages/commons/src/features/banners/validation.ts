import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';

import { BannerKind, BannerPlacement } from './types';

const BannerSourceSchema = yup
  .object({
    kind: yup.mixed().oneOf(Object.values(BannerKind)),
    url: yupTypes.nonEmptyString(),
  })
  .noUnknown();

export const BannerInputSchema = yup
  .object({
    id: yupTypes.uuid().defined().nullable(),
    slug: yupTypes.slug(),
    name: yupTypes.nonEmptyString(),
    priority: yup.number().integer(),
    enabled: yup.bool(),
    placement: yup.mixed().oneOf(Object.values(BannerPlacement)),
    source: BannerSourceSchema.clone(),
    link: yupTypes.https().nullable(true),
    extras: yup.object().nullable(true),
    regions: yup.array(yupTypes.node()),
    groups: yup.array(yupTypes.node()),
  })
  .strict(true)
  .noUnknown(true);
