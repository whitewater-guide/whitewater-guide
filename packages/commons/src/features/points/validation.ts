// tslint:disable-next-line
import * as Joi from 'joi';
import { POINames } from './POITypes';

export const CoordinateSchema = Joi.array().ordered(
  Joi.number().min(-180).max(180).required().label('longitude'), // longitude
  Joi.number().min(-90).max(90).required().label('latitude'), // latitude
  Joi.number().required().label('altitude'), // altitude
);

export const PointInputSchema = Joi.object().keys({
  id: Joi.string().guid().allow(null),
  name: Joi.string().min(3).max(100).allow(null),
  description: Joi.string().allow(null),
  coordinates: CoordinateSchema.required(),
  kind: Joi.any().allow(POINames),
});
