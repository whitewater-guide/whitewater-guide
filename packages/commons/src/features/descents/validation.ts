import { yupTypes } from '@whitewater-guide/validation';
import * as yup from 'yup';
import { DescentLevelInput, DescentInput } from './types';

export const DescentLevelInputSchema = yup
  .object<DescentLevelInput>({
    unit: yup
      .string()
      .max(20)
      .notRequired()
      .nullable(),
    value: yup.number().required(),
  })
  .strict(true)
  .noUnknown();

export const DescentInputSchema = yup
  .object<DescentInput>({
    id: yupTypes
      .uuid()
      .notRequired()
      .nullable(),
    sectionId: yupTypes
      .uuid()
      .required()
      .nullable(false),
    startedAt: yup.date().required() as any,
    duration: yup
      .number()
      .integer()
      .notRequired()
      .nullable(),
    level: DescentLevelInputSchema.clone()
      .notRequired()
      .nullable(),
    comment: yup
      .string()
      .notRequired()
      .nullable(),
    public: yup
      .bool()
      .notRequired()
      .nullable(),
  })
  .strict(true)
  .noUnknown();
