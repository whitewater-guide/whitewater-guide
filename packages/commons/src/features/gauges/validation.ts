import { baseStruct } from '../../utils/validation';
import { PointInputStruct } from '../points';

const GaugeInputFields = {
  id: baseStruct.union(['uuid', 'null']),
  name: 'nonEmptyString',
  code: 'nonEmptyVarchar',
  levelUnit: 'nonEmptyVarchar|null',
  flowUnit: 'nonEmptyVarchar|null',
  location: baseStruct.union([PointInputStruct, 'null']),
  requestParams: 'object|null',
  cron: baseStruct.union(['cron', 'null']),
  url: baseStruct.union(['url', 'null', baseStruct.literal('')]),
  source: 'node',
};

export const GaugeInputStruct = baseStruct.object(GaugeInputFields);

export const GaugeFormStruct = baseStruct.object({
  ...GaugeInputFields,
  requestParams: baseStruct.union(['jsonString', 'null']),
});
