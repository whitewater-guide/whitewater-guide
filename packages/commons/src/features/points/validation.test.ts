import { CoordinateSchema, PointInputSchema } from './validation';

import { PointInput } from './types';
import { createSafeValidator } from '@whitewater-guide/validation';

describe('loose coordinate schema', () => {
  const validator = createSafeValidator(CoordinateSchema);

  it('correct 2d', () => {
    expect(validator([120, 80])).toBeNull();
  });

  it('correct 3d', () => {
    expect(validator([120, 80, 22])).toBeNull();
  });

  it('undefined alt', () => {
    expect(validator([120, 80, undefined])).toBeNull();
  });

  it('null alt', () => {
    expect(validator([120, 80, null])).toBeNull();
  });

  it('error', () => {
    const errors = validator([220, 180]);
    expect(errors).toMatchSnapshot();
  });

  it('error when undefined', () => {
    const errors = validator(undefined);
    expect(errors).toMatchSnapshot();
  });

  it('wrong array length', () => {
    const errors = validator([3, 3, -11.5, 33]);
    expect(errors).not.toBeNull();
    expect(errors).toMatchSnapshot();
  });

  it('error when some are string', () => {
    const errors = validator([3, '3', 3]);
    expect(errors).not.toBeNull();
    expect(errors).toMatchSnapshot();
  });
});

describe('point input schema', () => {
  const validator = createSafeValidator(PointInputSchema);

  type TestValue = [string, PointInput];

  const correct: TestValue[] = [
    [
      'basic point',
      {
        id: null,
        name: null,
        description: null,
        coordinates: [125, 16, 11.1],
        kind: 'hazard',
      },
    ],
    [
      '2d point',
      {
        id: null,
        name: null,
        description: null,
        coordinates: [125, 16],
        kind: 'hazard',
      },
    ],
    [
      'point with undefined name and description',
      {
        id: null,
        coordinates: [125, 16, 11.1],
        kind: 'hazard',
      },
    ],
    [
      'point with empty name and description',
      {
        id: '7b6c1195-b74e-4560-b200-b41782237f3c',
        name: '',
        description: '',
        coordinates: [125, 16, 11.1],
        kind: 'hazard',
      },
    ],
    [
      'full point',
      {
        id: '7b6c1195-b74e-4560-b200-b41782237f3c',
        name: 'foo',
        description: 'bar',
        coordinates: [125, 16, 11.1],
        kind: 'hazard',
      },
    ],
  ];

  const incorrect: TestValue[] = [
    [
      'all',
      {
        id: 'foo',
        name: 'name',
        description: 'description',
        coordinates: [300, 300, 101],
        kind: 'bar',
      },
    ],
    [
      'lng only',
      {
        id: 'foo',
        name: 'name',
        description: 'description',
        coordinates: [30] as any,
        kind: 'put-in',
      },
    ],
    [
      'lat only',
      {
        id: 'foo',
        name: 'name',
        description: 'description',
        coordinates: [undefined, 30] as any,
        kind: 'put-in',
      },
    ],
    [
      'extra coordinate array item',
      {
        id: 'foo',
        name: 'name',
        description: 'description',
        coordinates: [30, 30, 30, 30] as any,
        kind: 'put-in',
      },
    ],
  ];

  it.each(correct)('should be valid for %s', (_, value) => {
    expect(validator(value)).toBeNull();
  });

  it.each(incorrect)('should be invalid for %s', (_, value) => {
    const result = validator(value);
    expect(result).not.toBeNull();
    expect(result).toMatchSnapshot();
  });
});
