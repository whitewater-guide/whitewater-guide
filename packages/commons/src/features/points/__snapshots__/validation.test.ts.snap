// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coordinate schema error in both 1`] = `
Object {
  "0": "yup:number.longitude",
  "1": "yup:number.latitude",
}
`;

exports[`coordinate schema error in lat 1`] = `
Object {
  "1": "yup:number.latitude",
}
`;

exports[`coordinate schema wrong array length (2) 1`] = `
Object {
  "2": "yup:number.altitude",
}
`;

exports[`coordinate schema wrong array length (4) 1`] = `
Object {
  "undefined": Object {
    "key": "yup:array.max",
    "options": Object {
      "max": 3,
      "path": "this",
    },
  },
}
`;

exports[`loose coordinate schema error 1`] = `
Object {
  "0": "yup:number.longitude",
  "1": "yup:number.latitude",
}
`;

exports[`loose coordinate schema wrong array length 1`] = `
Object {
  "undefined": Object {
    "key": "yup:array.max",
    "options": Object {
      "max": 3,
      "path": "this",
    },
  },
}
`;

exports[`point input schema should be invalid for all 1`] = `
Object {
  "coordinates": Array [
    "yup:number.longitude",
    "yup:number.latitude",
  ],
  "id": "yup:string.uuid",
  "kind": Object {
    "key": "yup:mixed.oneOf",
    "options": Object {
      "values": "put-in, put-in-alt, put-in-road, take-out, take-out-alt, take-out-road, waterfall, rapid, portage, playspot, hazard, river-campsite, wild-camping, paid-camping, gauge, hike-waypoint, bridge, dam, power-plant, kayak-shop, other",
    },
  },
}
`;

exports[`point input schema should be invalid for extra coordinate array item 1`] = `
Object {
  "coordinates": Object {
    "key": "yup:array.max",
    "options": Object {
      "max": 3,
      "path": "coordinates",
    },
  },
  "id": "yup:string.uuid",
}
`;

exports[`point input schema should be invalid for lat only 1`] = `
Object {
  "coordinates": Array [
    "yup:number.longitude",
    undefined,
    "yup:number.altitude",
  ],
  "id": "yup:string.uuid",
}
`;

exports[`point input schema should be invalid for lng only 1`] = `
Object {
  "coordinates": Array [
    undefined,
    "yup:number.latitude",
    "yup:number.altitude",
  ],
  "id": "yup:string.uuid",
}
`;

exports[`point input schema should be invalid for missing alt 1`] = `
Object {
  "coordinates": Array [
    undefined,
    undefined,
    "yup:number.altitude",
  ],
  "id": "yup:string.uuid",
}
`;
